### PR TITLE
Split the graphs to run with flash_attention on 1x

### PR DESCRIPTION
**Pull request by kalyanjk**
With flash attention enabled for larger batch sizes, recipe arc hbm memory size exceeds QueueComputeScal arc hbm memory.  Hence split the graph on 1x.

### CODE REVIEW

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -676,6 +677,9 @@ def forward(
         next_decoder_cache = () if not use_new_cache else None
 
         for layer_idx, decoder_layer in enumerate(self.layers):
+            if  torch.distributed.is_initialized() == False:
+                htcore.mark_step()
```
Comment (libinta) [inline optimum/habana/transformers/models/llama/modeling_llama.py:681]: @kalyanjk what's the impact for input/output not introduced oom? should we add an argument in text-generation from cmd line?

Comment (mandy-li) [reply optimum/habana/transformers/models/llama/modeling_llama.py:681]: @kalyanjk ,why only mark_step() for 1x?

Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py:681]: For 8x mark_step will be introduced through a collective call.

Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py:681]: > @kalyanjk what's the impact for input/output not introduced oom? should we add an argument in text-generation from cmd line?

The issue is not with oom. The real issue is recipe size being too large and also compilation time is too high.

**Comment by puneeshkhanna** [comment]
@kalyanjk  - we can abandon this PR. I have handled the change in #65. 
This also helps 8x inference.
I m checking 1x perf results too.
Further need to check finetuning script once.

**Comment by puneeshkhanna** [comment]
Wait we should not put mark step after the start of loop. Will create more graphs and perf is lower.

**Comment by kalyanjk** [comment]
> Wait we should not put mark step after the start of loop. Will create more graphs and perf is lower.
@puneeshkhanna 
On G3 we were seeing good perf with mark_step inside the for loop. With mark_step outside the for loop we are not able to run on single card.  This issue is also present in G2

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
     _gaudi_prepare_4d_causal_attention_mask,
 )
 
+import habana_frameworks.torch.core as htcore
```
Comment (msinnha1) [inline optimum/habana/transformers/models/llama/modeling_llama.py:26]: If you rebase to latest then this htcore import is not required, as it is part of PR#65

**Review by msinnha1 — COMMENTED** [review]
Verified the change and it is required for faster recipe compilation

**Review by dvarshney-habana — APPROVED** [review]

**Review by dvarshney-habana — APPROVED** [review]

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:696]: Please update as below
if lazy_mode and (torch.distributed.is_initialized() is False or torch.distributed.get_world_size() == 1):

**Comment by kalyanjk** [comment]
This PR solves the actual issue https://github.com/HabanaAI/optimum-habana-fork/pull/126

**Comment by astachowiczhabana** [comment]
Unmatched PR

**Pull request merged**